node.js

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// @ts-check
+
+/**
+ * This is the entrypoint on node-compatible CommonJS environments.
+ * `asyncLoad` will use `fs.readFile` to load the WASM module.
+ */
+
+const { readFileSync } = require("node:fs");
+const { readFile } = require("node:fs/promises");
+const path = require("node:path");
+const bindings = require("./pkg/matrix_sdk_crypto_wasm_bg.cjs");
+
+const filename = path.join(__dirname, "pkg/matrix_sdk_crypto_wasm_bg.wasm");
+
+// In node environments, we want to automatically load the WASM module
+// synchronously if the consumer did not call `initAsync`. To do so, we install
+// a `Proxy` that will intercept calls to the WASM module.
+bindings.__wbg_set_wasm(
+    new Proxy(
+        {},
+        {
+            get(_target, prop) {
+                const mod = loadModuleSync();
+                return initInstance(mod)[prop];
+            },
+        },
+    ),
+);
+
+/**
+ * Stores a promise which resolves to the WebAssembly module
+ * @type {Promise<WebAssembly.Module> | null}
+ */
+let modPromise = null;
+
+/**
+ * Tracks whether the module has been instantiated or not
+ * @type {boolean}
+ */
+let initialised = false;
+
+/**
+ * Loads the WASM module synchronously
+ *
+ * It will throw if there is an attempt to laod the module asynchronously running
+ *
+ * @returns {WebAssembly.Module}
+ */
+function loadModuleSync() {
+    if (modPromise) throw new Error("The WASM module is being loadded asynchronously but hasn't finished");
+    const bytes = readFileSync(filename);
+    return new WebAssembly.Module(bytes);
+}
+
+/**
+ * Loads the WASM module asynchronously
+ *
+ * @returns {Promise<WebAssembly.Module>}
+ */
+async function loadModule() {
+    const bytes = await readFile(filename);
+    return await WebAssembly.compile(bytes);
+}
+
+/**
+ * Initializes the WASM module and returns the exports from the WASM module.
+ *
+ * @param {WebAssembly.Module} mod
+ * @returns {typeof import("./pkg/matrix_sdk_crypto_wasm_bg.wasm.d")}
+ */
+function initInstance(mod) {
```

Comment:
```suggestion
function initInstance(mod) {
    if (initialised) {
        // This should be unreachable
        throw new Error("initInstance called twice");
    }
    
```

Reply:
This actually throws if you call `initAsync`, it loads in the background, and then init it synchronously before it finishes; but I think it is fine to throw because the consumer shouldn't do that

Reply:
> This actually throws if you call initAsync, it loads in the background, and then init it synchronously before it finishes;

Will it? In that case, wouldn't `modPromise` be set, so that the synchronous init will fail early with an exception?